### PR TITLE
Gh 752

### PR DESCRIPTION
**Pull request by viniciusccarvalho**
Adds caching capabilities to the schema registry client. Also pushing docs changes from previous PR

Resolves spring-cloud/spring-cloud-stream#752

### CODE REVIEW

File: spring-cloud-stream-reactive/dependency-reduced-pom.xml [hunk 1]
```diff
@@ -0,0 +1,144 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ilayaperumalg) [inline spring-cloud-stream-reactive/dependency-reduced-pom.xml]: I noticed this file keeps getting generated every time we run the build. May be we better add here. Is that the idea? /cc @mbogoevici

File: spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/CachingRegistryClient.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	private CacheManager cacheManager;
+
+	public CachingRegistryClient(SchemaRegistryClient delegate) {
+		this.delegate = delegate;
+	}
+
+	@Override
+	public SchemaRegistrationResponse register(String subject, String format, String schema) {
+		SchemaRegistrationResponse response = delegate.register(subject,format,schema);
+		cacheManager.getCache("refCache").put(response.getSchemaReference(),schema);
```
Comment (ilayaperumalg) [inline spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/CachingRegistryClient.java]: Would be better to make these `cache` names as constants.

File: spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientConfiguration.java [hunk 3]
```diff
@@ -25,17 +29,28 @@
...
 @EnableConfigurationProperties(SchemaRegistryClientProperties.class)
+@EnableCaching
 public class SchemaRegistryClientConfiguration {
 
+	@Autowired
+	private SchemaRegistryClientProperties schemaRegistryClientProperties;
+
 	@Bean
+	@ConditionalOnMissingBean(SchemaRegistryClient.class)
 	public SchemaRegistryClient schemaRegistryClient(SchemaRegistryClientProperties schemaRegistryClientProperties) {
```
Comment (ilayaperumalg) [inline spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientConfiguration.java]: Given this `SchemaRegistryClientProperties` is autowired and used inside this method, I believe we can remove the method argument.

File: spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientProperties.java [hunk 4]
```diff
@@ -28,6 +29,8 @@
 
 	private boolean enabled = true;
 
+	private boolean caching = true;
```
Comment (ilayaperumalg) [inline spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientProperties.java]: How is this property being used?

File: spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientProperties.java [hunk 5]
```diff
@@ -28,6 +29,8 @@
 
 	private boolean enabled = true;
```
Comment (ilayaperumalg) [inline spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientProperties.java]: I am not sure I understand how is the property `enabled` used as well?

File: spring-cloud-stream-schema/src/test/java/org/springframework/cloud/schema/client/SchemaClientTests.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+import org.springframework.test.context.junit4.SpringRunner;
+
+/**
+ * @author Vinicius Carvalho
+ */
+@RunWith(SpringRunner.class)
+@SpringBootTest(classes = SchemaClientTests.BootApplicationTest.class, webEnvironment = SpringBootTest.WebEnvironment.MOCK, properties = {"spring.cloud.stream.schemaRegistryClient.caching=true","debug=true"})
+@SpyBean(SchemaRegistryClient.class)
+public class SchemaClientTests {
+
```
Comment (ilayaperumalg) [inline spring-cloud-stream-schema/src/test/java/org/springframework/cloud/schema/client/SchemaClientTests.java]: We can avoid extra spaces in the classes in general.

File: spring-cloud-stream-reactive/dependency-reduced-pom.xml (hunk 1 above)
Comment (mbogoevici) [inline spring-cloud-stream-reactive/dependency-reduced-pom.xml]: We should add it to .gitignore  and not check it in.

File: spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientConfiguration.java (hunk 3 above)
Comment (mbogoevici) [inline spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientConfiguration.java]: `@ConditionalOnMissingBean` is unsafe for use outside an autoconfiguration in general. (We make an exception for it in `BindingServiceConfiguration` and that is exclusively for hierarchies).

File: spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientProperties.java (hunk 4 above)
Comment (viniciusccarvalho) [inline spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientProperties.java]: I don't think it is, I can't really remember why we have it, I'll leave this one out, fixing all other comments on a new commit.

File: spring-cloud-stream-reactive/dependency-reduced-pom.xml (hunk 1 above)
Comment (mbogoevici) [inline spring-cloud-stream-reactive/dependency-reduced-pom.xml]: We also need to remove this file from the commit.

File: spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientConfiguration.java [hunk 7]
```diff
@@ -16,7 +16,11 @@
 
 package org.springframework.cloud.stream.schema.client.config;
 
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
```
Comment (mbogoevici) [inline spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientConfiguration.java]: Unused import.

File: spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientProperties.java [hunk 8]
```diff
@@ -43,4 +45,6 @@ public boolean isEnabled() {
 	public void setEnabled(boolean enabled) {
```
Comment (mbogoevici) [inline spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/config/SchemaRegistryClientProperties.java]: We can remove this property as it is not used in code. The usage within `@ConditionalOnProperties` depends on the presence (or absence) of the property in the environment.

File: spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/CachingRegistryClient.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * @author Vinicius Carvalho
+ */
+public class CachingRegistryClient implements SchemaRegistryClient {
+
+	private SchemaRegistryClient delegate;
+
+	protected static final String ID_CACHE = "idCache";
+	protected static final String REF_CACHE = "refCache";
```
Comment (mbogoevici) [inline spring-cloud-stream-schema/src/main/java/org/springframework/cloud/stream/schema/client/CachingRegistryClient.java]: I think the names of the caches should be more Stream-specific (easy to clash with someone's code). Maybe we can consider making them configurable in the future.

File: spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc [hunk 10]
```diff
@@ -1439,6 +1451,13 @@ A client for the Spring Cloud Stream schema registry can be configured using the
   }
 ----
 
+[NOTE]
+====
+The default client caches all responses locally to avoid unnecessary network calls. It leverages Spring caching abstraction, so you can provide
+an external cache provider such as eh-cache or hazelcast if you want to have distributed cache control and invalidation. Schemas are append only entities
+but deletes may occur, make sure you configure your external cache provider to invalidate the cache from time to time to avoid synchronization issues.
+====
```
Comment (mbogoevici) [inline spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc]: Can we detail a little more how that works, as well as the names of the two cache regions that we use?

**Comment by mbogoevici** [comment]
Closing in favour of https://github.com/spring-cloud/spring-cloud-stream/pull/835

**Pull request closed without merging**